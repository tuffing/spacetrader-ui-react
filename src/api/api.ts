// @todo implement some error states and types

// should do a dev / staging etc check here. but spacetraders only has the one url endpoint
export const url = 'https://api.spacetraders.io';

export enum Status {
	idle = 'idle',
	loading = 'loading',
	succeeded = 'succeeded',
	failed = 'failed'
}

export const performRequest = async (
	method: 'POST' | 'GET',
	endpoint: string,
	headers?: HeadersInit,
	stringifiedData?: string
) => {
	const res: RequestInit = {
		method: method,
		cache: 'no-cache',
		credentials: 'same-origin', // include, *same-origin, omit
		headers: {
			'Content-Type': 'application/json',
			...headers
		}
	};
	if (stringifiedData) res.body = stringifiedData;

	const response = await fetch(`${url}/${endpoint}`, res);

	return response.json();
};
