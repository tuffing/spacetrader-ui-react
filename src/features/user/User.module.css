.username {
	font-weight: bold;
}

.infoPanel {
	font-size: 16px;
}

.infoPanel div > div {
	margin-right: 10px;
}

.infoPanel:last-child {
	margin-right: 0;
}
