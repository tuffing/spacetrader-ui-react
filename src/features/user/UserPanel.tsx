import React from 'react';
import { RootState } from '../../app/store';
import { useSelector } from 'react-redux';
import userStyles from './User.module.css';
import globalStyles from '../../Styles.module.css';

export function UserPanel() {
	const username = useSelector((state: RootState) => state.userReducer.username);
	const credits = useSelector((state: RootState) => state.userReducer.credits);

	return (
		<div className={userStyles.infoPanel}>
			<div className={globalStyles.row}>
				<div className={userStyles.username}>{username}</div>
				<div className={userStyles.userData} aria-label="Credits">
					Credits {credits}
				</div>
			</div>
		</div>
	);
}
