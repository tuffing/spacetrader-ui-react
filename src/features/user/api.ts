// @TODO turn into a proper api library.
export const authCall = async (username: string) => {
	const response = await fetch(`https://api.spacetraders.io/users/${username}/token`, {
		method: 'POST',
		//mode: 'no-cors', // no-cors, *cors, same-origin
		cache: 'no-cache',
		credentials: 'same-origin', // include, *same-origin, omit
		headers: {
			'Content-Type': 'application/json',
		},
	});

	return response.json();
};

export interface AuthBody {
	token: string;
	user: {
		id: string;
		username: string;
		picture: string | null;
		email: string | null;
		credits: number;
		createdAt: string;
		updatedAt: string;
	};
}
