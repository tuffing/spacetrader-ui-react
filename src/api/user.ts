import { url } from './api';

export interface AuthTokenBody {
	token: string;
	user: {
		id: string;
		username: string;
		picture: string | null;
		email: string | null;
		credits: number;
		createdAt: string;
		updatedAt: string;
	};
}

export interface AuthTokenBodyFail {
	error: {
		message: string;
		code: number;
	};
}

export const requestToken = async (username: string) => {
	const response = await fetch(`${url}/users/${username}/token`, {
		method: 'POST',
		cache: 'no-cache',
		credentials: 'same-origin', // include, *same-origin, omit
		headers: {
			'Content-Type': 'application/json',
		},
	});

	return response.json();
};
