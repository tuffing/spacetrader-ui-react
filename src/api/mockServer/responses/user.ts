import { setUserName } from '../../../features/user/userSlice';
import { AuthTokenBody, AuthTokenBodyFail } from '../../user';

export enum AuthFailParams {
	usernameTaken = 'user-taken'
}

export const authTokenSuccessBody: AuthTokenBody = {
	token: '12345678-1234-1234-1234-12345678912',
	user: {
		id: 'cklxd8vpr301164iw89x8186uts',
		username: 'abcdef',
		picture: null,
		email: null,
		credits: 15,
		createdAt: '2021-03-06T06:49:04.144Z',
		updatedAt: '2021-03-06T06:49:04.143Z'
	}
};

export const authTokenErrorUserTakenBody: AuthTokenBodyFail = {
	error: {
		message: 'Username has already been claimed.',
		code: 40901
	}
};
