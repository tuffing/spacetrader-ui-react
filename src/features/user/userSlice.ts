import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { Status } from '../../api/api';
import { requestToken } from '../../api/user';

export interface UserState {
	username: string;
	credits: number;
	token?: string;
	status: Status;
	error?: string;
}

export const initialState: UserState = {
	username: '',
	credits: 0,
	status: Status.idle
};

// fetch the user token
export const authenticate = createAsyncThunk('user/authenticate', async (username: string) => {
	return requestToken(username);
});

export const userSlice = createSlice({
	name: 'user',
	initialState,
	reducers: {
		// Use the PayloadAction type to declare the contents of `action.payload`
		setUserName: (state, action: PayloadAction<string>) => {
			state.username = action.payload;
		},
		setAccessToken: (state, action: PayloadAction<string>) => {
			state.token = action.payload;
		},
		setCredits: (state, action: PayloadAction<number>) => {
			state.credits = action.payload;
		},
		setStatus: (state, action: PayloadAction<Status>) => {
			state.status = action.payload;
		}
	},
	extraReducers: (builder) => {
		builder
			.addCase(authenticate.pending, (state, action) => {
				state.status = Status.loading;
			})
			.addCase(authenticate.fulfilled, (state, action) => {
				if ('error' in action.payload) {
					//validation errors etc return as 200s. So validation here..
					state.error = action.payload.error.message;
					state.status = Status.failed;
					return;
				}

				state.token = action.payload.token;
				state.credits = action.payload.user.credits;
				state.username = action.payload.user.username;

				state.status = Status.idle;
			})
			.addCase(authenticate.rejected, (state, action) => {
				state.status = Status.failed;
				if ('message' in action.error) state.error = action.error.message;
				else state.error = 'undefined error';
			});
	}
});

export const { setUserName, setAccessToken, setCredits, setStatus } = userSlice.actions;

export default userSlice.reducer;
