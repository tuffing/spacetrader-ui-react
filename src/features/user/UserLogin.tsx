import React, { useState } from 'react';
import { RootState } from '../../app/store';
import { Status } from '../../api/api';
import { useSelector, useDispatch } from 'react-redux';
import { authenticate } from './userSlice';
import userStyles from './User.module.css';
import styles from '../../Styles.module.css';

export default function UserLogin() {
	const token = useSelector((state: RootState) => state.userReducer.token);
	const error = useSelector((state: RootState) => state.userReducer.error);
	const status = useSelector((state: RootState) => state.userReducer.status);
	const dispatch = useDispatch();
	const [userName, setUserName] = useState('');

	return (
		<div>
			<div>
				Auth token: <span data-testid="token">{token ?? '[not set]'}</span>
			</div>
			<div>
				Errors: <span>{error ?? '[none]'}</span>
			</div>

			<div className={styles.row}>
				<input
					data-testid="UserName"
					className={styles.textbox}
					aria-label="User name"
					value={userName}
					onChange={(e) => setUserName(e.target.value)}
				/>
				<button
					data-testid="LoginSubmit"
					className={styles.button}
					disabled={!userName || status == Status.loading}
					onClick={() => dispatch(authenticate(userName))}
				>
					Login
				</button>
			</div>
			<div>Auth status: {status}</div>
			<div>Username: {userName}</div>
		</div>
	);
}
