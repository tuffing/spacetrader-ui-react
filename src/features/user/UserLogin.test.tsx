import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import { store } from '../../app/store';
import UserLogin from './UserLogin';
import { setStatus, setAccessToken } from './userSlice';
import { Status } from '../../api/api';
import { AuthTokenSuccessBody, AuthTokenErrorUserTakenBody, AuthFailParams } from '../../api/mockServer/responses/user';

describe('User Login Component', () => {
	beforeEach(() => {
		store.dispatch(setStatus(Status.idle));
		store.dispatch(setAccessToken(''));
	});

	it('User snapshot', () => {
		//@todo switch with enzyme or similar?
		const login = render(
			<Provider store={store}>
				<UserLogin />
			</Provider>
		);

		expect(login).toMatchSnapshot();
	});

	it('Login Disabled if username empty', () => {
		const login = render(
			<Provider store={store}>
				<UserLogin />
			</Provider>
		);

		expect(screen.getByTestId('LoginSubmit')).toHaveAttribute('disabled');
		userEvent.type(screen.getByTestId('UserName'), 'testUser');
		expect(screen.getByTestId('UserName')).toHaveValue('testUser');
		expect(screen.getByTestId('LoginSubmit')).not.toHaveAttribute('disabled');
	});

	it('Login disabled while checking auth', () => {
		store.dispatch(setStatus(Status.loading));
		const login = render(
			<Provider store={store}>
				<UserLogin />
			</Provider>
		);
		expect(screen.getByTestId('LoginSubmit')).toHaveAttribute('disabled');
		userEvent.type(screen.getByTestId('UserName'), 'testUser');
		//should stil be disabled
		expect(screen.getByTestId('LoginSubmit')).toHaveAttribute('disabled');
	});

	it('Login Button attempts login', async () => {
		const login = render(
			<Provider store={store}>
				<UserLogin />
			</Provider>
		);

		userEvent.type(screen.getByTestId('UserName'), 'testUser');
		userEvent.click(screen.getByTestId('LoginSubmit'));

		// @todo replace with something more logical.  in reality the token won't (shouldnt be!) be rendered
		expect(await screen.findByText(AuthTokenSuccessBody.token)).toBeInTheDocument();
	});

	it('Login Button attempts login but username is taken', async () => {
		const login = render(
			<Provider store={store}>
				<UserLogin />
			</Provider>
		);

		userEvent.type(screen.getByTestId('UserName'), AuthFailParams.usernameTaken.toString());
		userEvent.click(screen.getByTestId('LoginSubmit'));
		expect(await screen.findByText(AuthTokenErrorUserTakenBody.error.message)).toBeInTheDocument();
	});

	// @todo connection error?
});
