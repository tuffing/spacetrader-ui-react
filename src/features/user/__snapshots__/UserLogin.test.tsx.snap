// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Login Component User snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          Auth token: 
          <span
            data-testid="token"
          />
        </div>
        <div>
          Errors: 
          <span>
            [none]
          </span>
        </div>
        <div
          class="row"
        >
          <input
            aria-label="User name"
            class="textbox"
            data-testid="UserName"
            value=""
          />
          <button
            class="button"
            data-testid="LoginSubmit"
            disabled=""
          >
            Login
          </button>
        </div>
        <div>
          Auth status: 
          idle
        </div>
        <div>
          Username: 
          
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        Auth token: 
        <span
          data-testid="token"
        />
      </div>
      <div>
        Errors: 
        <span>
          [none]
        </span>
      </div>
      <div
        class="row"
      >
        <input
          aria-label="User name"
          class="textbox"
          data-testid="UserName"
          value=""
        />
        <button
          class="button"
          data-testid="LoginSubmit"
          disabled=""
        >
          Login
        </button>
      </div>
      <div>
        Auth status: 
        idle
      </div>
      <div>
        Username: 
        
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
